src/transport/icmp4.rs

```diff
@@ -0,0 +1,243 @@
+use std::{slice::from_raw_parts, fmt::Display};
+
+use super::super::*;
+
+
+/**
+ * Generic ICMP Header - applies to IcmpV4Header and IcmpV6Header
+ */
+
+pub trait IcmpHeader {
+    fn get_raw_code(&self) -> u8;
+    fn get_raw_type(&self) -> u8;
+    fn get_checksum(&self) -> u16;
+    /// this u32 data has different meaning by the Icmp Message Type
+    /// For example, 
+    /// * with Echo Request/Reply, it's (seq: u16, id: u16)
+    /// * with Dest unreachable, it's reserved
+    fn get_type_dependent_data(&self) -> u32;
+}
+
+
+// Make sure all Icmp4 + Icmp6 headers have a default Display
+impl Display for dyn IcmpHeader {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "type={} code={}", self.get_raw_type(), self.get_raw_code())
+    }
+}
```

Comment:
I am not sure if it is worth it to try to build an abstraction between IcmpV4 & IcmpV6.

I would leave the trait out in the first implementation and let the user abstract over the two protocols on their own if they need it.